### PR TITLE
Themes: Use dynamic theme filters data

### PR DESCRIPTION
**Pull request by seear**
No visual changes.

Use data fetched from /theme-filters API endpoint for the Theme Showcase filter search. Utilises new selectors added in #13928 and #14060, and removes the static data and library `theme-filters.js`

**To Test**
* Check that the theme filters in the search box at /themes work as before
* Check that urls containing filters and verticals work as before, including logged-out
For example:
http://calypso.localhost:3000/themes/business
http://calypso.localhost:3000/themes/filter/one-column+classic-menu+clean+artwork

### CODE REVIEW

**Comment by folletto** [comment]
Tested in Chrome, a few set links (see above) and clicking around. Also tested searching and logged out. All seems working. 👍

**Comment by seear** [comment]
Removing the 'needs review' tag from this for now whilst I look at splitting the `theme-filters` lib into a couple of selectors and a collection of util functions.

**Comment by ockham** [comment]
Played a bit with this. I think we can tackle this mostly at consumer level, where normally `getThemeFilterTerm` and `findThemeFilterTerm` will be sufficient. I added `isValidThemeFilterTerm` to the mix which uses `findThemeFilterTerm` (which is modified to return `null` if the slug is ambiguous).

**Comment by seear** [comment]
Thanks for taking a look @ockham!

I had begun to formulate a different approach, where all the comsumer level stuff would remain untouched (apart from probably the validation middlewares) to prevent breaking anything, then splitting `theme-filters.js` into a couple of new selectors:
* `getTaxonomies()`
* `getTermTable()`

And turning all the other funcs into utils that would accept the data from either of the above selectors as a param, instead of calling them as it does now. It would spoil the functional nature of the code a bit, but should only need a couple of tweaks. I was thinking that it would probably be a fairly minimal PR (likely a brand new one).

I don't feel strongly about it, so if you want to continue with this PR, feel free to carry on :)

**Comment by seear** [comment]
Gone through the PR again. Seems like both approaches would end up at the same place. Interesting that you've chosen to touch the consumer stuff first. I think you'll get better consumer code this way, but with a greater risk of breaking something.

**Comment by seear** [comment]
Rebased

File: client/state/selectors/get-theme-filter-terms-table.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		const termTable = {};
+		const taxonomies = mapValues( getThemeFilters( state ), keys );
+		forIn( taxonomies, ( terms, taxonomy ) => {
+			terms.forEach( ( term ) => {
+				const key = isAmbiguousThemeFilterTerm( state, term ) ? `${ taxonomy }:${ term }` : term;
+				termTable[ key ] = taxonomy;
+			} );
+		} );
+		return termTable;
+	}
```
Comment (ockham) [inline client/state/selectors/get-theme-filter-terms-table.js]: We need to supply a second arg to `createSelector` to actually memoize.

File: client/state/selectors/prepend-theme-filter-keys.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * "taxonomy:term taxonomy:term" search-box format.
+ *
+ * @param {Object} state Global state tree
+ * @param {string} terms Space or + separated list of filter terms
+ * @return {string} Complete taxonomy:term filter string, or empty string if term is not valid
+ */
+export default function prependThemeFilterKeys( state, terms ) {
+	const getFilter = curry( getThemeFilterStringFromTerm )( state );
+	if ( terms ) {
+		return terms.split( /[+\s]/ ).map( getFilter ).join( ' ' ) + ' ';
```
Comment (ockham) [inline client/state/selectors/prepend-theme-filter-keys.js]: Can't we just inline `( term => getThemeFilterStringFromTerm( state, term ) )` here? I'd find that more explicit / easier to read... (same for a bunch of other 🍛s in this PR)

File: client/state/selectors/get-theme-filter-terms-table.js (hunk 1 above)
Comment (ockham) [inline client/state/selectors/get-theme-filter-terms-table.js]: (I'll do that on #14060.)

File: client/state/selectors/prepend-theme-filter-keys.js (hunk 2 above)
Comment (seear) [inline client/state/selectors/prepend-theme-filter-keys.js]: Sure! I was using curries to use selectors with the original code, but we can factor them out now.

**Pull request merged**